nx/source/runtime/devices/socket.c

```diff
@@ -0,0 +1,1550 @@
+
+static int _socketOpen(struct _reent *r, void *fdptr, const char *path, int flags, int mode);
+static int _socketClose(struct _reent *r, void *fdptr);
+static ssize_t _socketWrite(struct _reent *r, void *fdptr, const char *buf, size_t count);
+static ssize_t _socketRead(struct _reent *r, void *fdptr, char *buf, size_t count);
+
+static const devoptab_t g_socketDevoptab = {
+    .name = "soc",
+    .structSize   = sizeof(int),
+    .open_r       = _socketOpen,
+    .close_r      = _socketClose,
+    .write_r      = _socketWrite,
+    .read_r       = _socketRead,
+    .seek_r       = NULL,
+    .fstat_r      = NULL,
+    .stat_r       = NULL,
+    .link_r       = NULL,
+    .unlink_r     = NULL,
+    .chdir_r      = NULL,
+    .rename_r     = NULL,
+    .mkdir_r      = NULL,
+    .dirStateSize = 0,
+    .diropen_r    = NULL,
+    .dirreset_r   = NULL,
+    .dirnext_r    = NULL,
+    .dirclose_r   = NULL,
+    .statvfs_r    = NULL,
+    .ftruncate_r  = NULL,
+    .fsync_r      = NULL,
+    .deviceData   = 0,
+    .chmod_r      = NULL,
+    .fchmod_r     = NULL,
+    .rmdir_r      = NULL,
+};
+
+static const SocketInitConfig g_defaultSocketInitConfig = {
+    .bsdsockets_version = 1,
+
+    .tcp_tx_buf_size        = 0x8000,
+    .tcp_rx_buf_size        = 0x10000,
+    .tcp_tx_buf_max_size    = 0x40000,
+    .tcp_rx_buf_max_size    = 0x40000,
+
+    .udp_tx_buf_size = 0x2400,
+    .udp_rx_buf_size = 0xA500,
+
+    .sb_efficiency = 4,
+
+    .serialized_out_addrinfos_max_size  = 0x1000,
+    .serialized_out_hostent_max_size    = 0x200,
+    .bypass_nsd                         = false,
+    .dns_timeout                        = 0,
+};
+
+const SocketInitConfig *socketGetDefaultInitConfig(void) {
+    return &g_defaultSocketInitConfig;
+}
+
+Result socketInitialize(const SocketInitConfig *config) {
+    Result ret = 0;
+    BsdInitConfig bcfg = {
+        .version = config->bsdsockets_version,
+
+        .tcp_tx_buf_size        = config->tcp_tx_buf_size,
+        .tcp_rx_buf_size        = config->tcp_rx_buf_size,
+        .tcp_tx_buf_max_size    = config->tcp_tx_buf_max_size,
+        .tcp_rx_buf_max_size    = config->tcp_rx_buf_max_size,
+
+        .udp_tx_buf_size = config->udp_tx_buf_size,
+        .udp_rx_buf_size = config->udp_rx_buf_size,
+
+        .sb_efficiency = config->sb_efficiency,
+    };
+
+    SfdnsresConfig sfdnsresConfig = {
+        .serialized_out_addrinfos_max_size  = config->serialized_out_addrinfos_max_size,
+        .serialized_out_hostent_max_size    = config->serialized_out_hostent_max_size,
+        .bypass_nsd                         = config->bypass_nsd,
+        .timeout                            = config->dns_timeout,
+    };
+
+    int dev = FindDevice("soc:");
+    if(dev != -1)
+        return MAKERESULT(Module_Libnx, LibnxError_AlreadyInitialized);
+
+    ret = bsdInitialize(&bcfg);
+    if(R_SUCCEEDED(ret))
+        dev = AddDevice(&g_socketDevoptab);
+
```

Comment:
Maybe I'm wrong but doesn't do proper cleanup on ret error here

Reply:
The functions returns early a few lines above if dev != -1, thus dev == -1 in this block, so this is fine.